presto-main/src/main/java/com/facebook/presto/operator/aggregation/KeySetAndValues.java

```diff
@@ -0,0 +1,220 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.type.Type;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.facebook.presto.common.type.BigintType.BIGINT;
+import static com.facebook.presto.common.type.DoubleType.DOUBLE;
+import static com.facebook.presto.common.type.IntegerType.INTEGER;
+import static com.facebook.presto.common.type.RealType.REAL;
+import static com.facebook.presto.type.TypeUtils.expectedValueSize;
+import static java.lang.Double.longBitsToDouble;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class KeySetAndValues
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(KeySetAndValues.class).instanceSize();
+    private static final int EXPECTED_ENTRIES = 10;
+    private static final int EXPECTED_ENTRY_SIZE = 16;
+
+    private final TypedSet keySet;
+    private final Type keyType;
+
+    private final BlockBuilder valueBlockBuilder;
+    private final Type valueType;
+
+    public KeySetAndValues(Type keyType, Type valueType)
+    {
+        this(keyType,
+                valueType,
+                new TypedSet(keyType, EXPECTED_ENTRIES, "MAP_UNION_SUM"),
+                valueType.createBlockBuilder(null, EXPECTED_ENTRIES, expectedValueSize(valueType, EXPECTED_ENTRY_SIZE)));
+    }
+
+    public KeySetAndValues(Type keyType, Type valueType, TypedSet keySet, BlockBuilder valueBlockBuilder)
+    {
+        this.keyType = requireNonNull(keyType, "keyType is null");
+        this.valueType = requireNonNull(valueType, "valueType is null");
+        this.keySet = keySet;
+        this.valueBlockBuilder = valueBlockBuilder;
+    }
+
+    public Type getKeyType()
+    {
+        return keyType;
+    }
+
+    public Block getKeys()
+    {
+        return keySet.getBlock();
+    }
+
+    public Block getValues()
+    {
+        return valueBlockBuilder.build();
+    }
+
+    public void serialize(BlockBuilder out)
+    {
+        BlockBuilder mapBlockBuilder = out.beginBlockEntry();
+        Block keyBlock = keySet.getBlock();
+        for (int i = 0; i < keyBlock.getPositionCount(); i++) {
+            keyType.appendTo(keyBlock, i, mapBlockBuilder);
+            valueType.appendTo(valueBlockBuilder, i, mapBlockBuilder);
+        }
+        out.closeEntry();
+    }
+
+    public long estimatedInMemorySize()
+    {
+        long size = INSTANCE_SIZE;
```

Comment:
Just use `return INSTANCE_SIZE + ... + ...`. We normally don't create locals in this situation.

Reply:
Done